services/Skyra.Notifications/Services/YoutubeService.cs

```diff
@@ -0,0 +1,147 @@
+using Skyra.Database;
+using Skyra.Grpc.Services;
+using Skyra.Grpc.Services.Shared;
+using Skyra.Notifications.Models;
+using Action = Skyra.Grpc.Services.Action;
+using Status = Skyra.Grpc.Services.Shared.Status;
+using YoutubeServiceBase = Skyra.Grpc.Services.YoutubeSubscription.YoutubeSubscriptionBase;
+
+namespace Skyra.Notifications.Services
+{
+	public class YoutubeService : YoutubeServiceBase
+	{
+		private readonly IDatabase _database;
+		private readonly ConcurrentQueue<Notification> _notificationQueue;
+		private readonly SubscriptionManager _subscriptionManager;
+		private readonly ILogger<YoutubeService> _logger;
+
+		public YoutubeService(IDatabase database, SubscriptionManager subscriptionManager, ConcurrentQueue<Notification> notificationQueue, ILogger<YoutubeService> logger)
+		{
+			_database = database;
+			_subscriptionManager = subscriptionManager;
+			_notificationQueue = notificationQueue;
+			_logger = logger;
+		}
+
+		public override async Task<SubscriptionResponse> GetSubscriptions(Empty request, ServerCallContext _)
+		{
+			var subscriptions = await _database.GetSubscriptionsAsync();
+
+			// lord I hate google sometimes
+
+			var response = new SubscriptionResponse();
+			foreach (var sub in subscriptions.Value)
+			{
+				var subscription = new Subscription
+				{
+					ChannelId = sub.Id
+				};
+
+				subscription.GuildIds.AddRange(sub.GuildIds);
+
+				response.Subscriptions.Add(subscription);
+			}
+
+			return response;
+		}
+
+		public override Task<Result> ManageSubscription(SubscriptionManageRequest request, ServerCallContext _)
+		{
+			return request.Type switch
+			{
+				Action.Subscribe => HandleSubscription(request.ChannelUrl, request.GuildId, request.NotificationMessage, request.GuildChannelId),
+				Action.Unsubscribe => HandleUnsubscription(request.ChannelUrl, request.GuildId),
+				_ => throw new ArgumentOutOfRangeException()
+			};
+		}
+
+		public override async Task<Result> UpdateSubscriptionSettings(NotificationSettingsUpdate request, ServerCallContext context)
+		{
+			var updated = await _database.UpdateYoutubeSubscriptionSettingsAsync(request.GuildId, request.Message, request.Channel)
+			return new Result
+			{
+				Status = updated.Success ? Status.Success : Status.Failed
+			};
+		}
+
+		public override async Task SubscriptionNotifications(Empty request, IServerStreamWriter<SubscriptionNotification> responseStream, ServerCallContext context)
+		{
+			while (true)
+			{
+				if (!_notificationQueue.TryDequeue(out var notification))
+				{
+					continue;
+				}
+
+				_logger.LogInformation("Sending notification {@Notification}", notification);
+
+				await using var database = new SkyraDatabase(new SkyraDbContext(), new NullLogger<SkyraDatabase>());
+
+				var notificationSubscription = await database.GetSubscriptionAsync(notification.ChannelId);
+
+				var channels = new NotificationChannel[notificationSubscription.Value.GuildIds.Length];
```

Comment:
currently not checking here if no guilds are subscribed (though technically it's handled by the for loop). Will need to check and remove the entry from the database all together

Reply:
Is this resolved yet?